File: app/src/androidTest/java/com/example/popularmovies/data/TestDb.java
package com.example.popularmovies.data;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.HashSet;

import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.greaterThan;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class TestDb {
    public static final String LOG_TAG = TestDb.class.getSimpleName();

    private Context mContext = InstrumentationRegistry.getTargetContext();

    @Before
    public void deleteDatabase() {
        mContext.deleteDatabase(MoviesDbHelper.DATABASE_NAME);
    }

    @Test
    public void testCreateDb() {
        final HashSet<String> tableNameHashSet = new HashSet<>();
        tableNameHashSet.add(MoviesContract.MovieEntry.TABLE_NAME);
        tableNameHashSet.add(MoviesContract.ListEntry.TABLE_NAME);
        tableNameHashSet.add(MoviesContract.MovieListEntry.TABLE_NAME);

        deleteDatabase();
        SQLiteDatabase db = new MoviesDbHelper(mContext).getWritableDatabase();
        assertEquals(true, db.isOpen());

        Cursor c = db.rawQuery("SELECT name FROM sqlite_master WHERE type='table'", null);
        assertTrue("Error: This means that the database has not been created correctly",
                c.moveToFirst());

        // verify that the tables have been created
        do {
            tableNameHashSet.remove(c.getString(0));
        } while( c.moveToNext() );
        c.close();
        assertTrue("Error: Your database was created without movie, list and movie_list entry tables",
                tableNameHashSet.isEmpty());

        // now, do our tables contain the correct columns?
        c = db.rawQuery("PRAGMA table_info(" + MoviesContract.MovieEntry.TABLE_NAME + ")",
                null);

        assertTrue("Error: Unable to query the database for movie table information.",
                c.moveToFirst());

        // Build a HashSet of all of the column names we want to look for
        final HashSet<String> movieColumnHashSet = new HashSet<>();
        movieColumnHashSet.add(MoviesContract.MovieEntry._ID);
        movieColumnHashSet.add(MoviesContract.MovieEntry.COLUMN_TITLE);
        movieColumnHashSet.add(MoviesContract.MovieEntry.COLUMN_OVERVIEW);
        movieColumnHashSet.add(MoviesContract.MovieEntry.COLUMN_POSTER_PATH);
        movieColumnHashSet.add(MoviesContract.MovieEntry.COLUMN_RELEASE_DATE);
        movieColumnHashSet.add(MoviesContract.MovieEntry.COLUMN_ADULT);
        movieColumnHashSet.add(MoviesContract.MovieEntry.COLUMN_VOTE_AVG);

        int columnNameIndex = c.getColumnIndex("name");
        do {
            String columnName = c.getString(columnNameIndex);
            movieColumnHashSet.remove(columnName);
        } while(c.moveToNext());
        c.close();
        // if this fails, it means that your database doesn't contain all of the required location
        // entry columns
        assertTrue("Error: The database doesn't contain all of the required movie entry columns",
                movieColumnHashSet.isEmpty());

        c = db.rawQuery("PRAGMA table_info(" + MoviesContract.ListEntry.TABLE_NAME + ")",
                null);

        assertTrue("Error: Unable to query the database for list table information.",
                c.moveToFirst());

        // Build a HashSet of all of the column names we want to look for
        final HashSet<String> listColumnHashSet = new HashSet<>();
        listColumnHashSet.add(MoviesContract.ListEntry._ID);
        listColumnHashSet.add(MoviesContract.ListEntry.COLUMN_SELECTION);
        listColumnHashSet.add(MoviesContract.ListEntry.COLUMN_DATE_FETCHED);

        columnNameIndex = c.getColumnIndex("name");
        do {
            String columnName = c.getString(columnNameIndex);
            listColumnHashSet.remove(columnName);
        } while(c.moveToNext());
        c.close();
        // if this fails, it means that your database doesn't contain all of the required location
        // entry columns
        assertTrue("Error: The database doesn't contain all of the required list entry columns",
                listColumnHashSet.isEmpty());

        c = db.rawQuery("PRAGMA table_info(" + MoviesContract.MovieListEntry.TABLE_NAME + ")",
                null);

        assertTrue("Error: Unable to query the database for list table information.",
                c.moveToFirst());

        // Build a HashSet of all of the column names we want to look for
        final HashSet<String> movieListColumnHashSet = new HashSet<>();
        movieListColumnHashSet.add(MoviesContract.MovieListEntry._ID);
        movieListColumnHashSet.add(MoviesContract.MovieListEntry.COLUMN_LIST_KEY);
        movieListColumnHashSet.add(MoviesContract.MovieListEntry.COLUMN_MOVIE_KEY);
        movieListColumnHashSet.add(MoviesContract.MovieListEntry.COLUMN_ORDER);

        columnNameIndex = c.getColumnIndex("name");
        do {
            String columnName = c.getString(columnNameIndex);
            movieListColumnHashSet.remove(columnName);
        } while(c.moveToNext());
        c.close();
        // if this fails, it means that your database doesn't contain all of the required location
        // entry columns
        assertTrue("Error: The database doesn't contain all of the required movie list entry columns",
                movieListColumnHashSet.isEmpty());

        db.close();
    }

    @Test
    public void testMovieTable() {
        MoviesDbHelper dbHelper = new MoviesDbHelper(mContext);
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues movie = TestUtilities.createEmpireStrikesBackValues();
        Long movieId;
        movieId = db.insert(MoviesContract.MovieEntry.TABLE_NAME, null, movie);
        assertThat(movieId, equalTo(TestUtilities.TEST_MOVIE2_ID));

        // Fourth Step: Query the database and receive a Cursor back
        // A cursor is your primary interface to the query results.
        Cursor cursor = db.query(
                MoviesContract.MovieEntry.TABLE_NAME,  // Table to Query
                null, // all columns
                null, // Columns for the "where" clause
                null, // Values for the "where" clause
                null, // columns to group by
                null, // columns to filter by row groups
                null // sort order
        );

        assertTrue( "Error: No Records returned from movie query", cursor.moveToFirst() );

        TestUtilities.validateCurrentRecord("Error: Movie Query Validation Failed",
                cursor, movie);

        // Move the cursor to demonstrate that there is only one record in the database
        assertFalse( "Error: More than one record returned from movie query",
                cursor.moveToNext() );

        // Close Cursor and Database
        cursor.close();
        db.close();
    }

    @Test
    public void testListTable() {
        MoviesDbHelper dbHelper = new MoviesDbHelper(mContext);
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues list = TestUtilities.createMostPopularListValues();
        Long listId;
        listId = db.insert(MoviesContract.ListEntry.TABLE_NAME, null, list);
        assertThat(listId, greaterThan(0L));

        // Fourth Step: Query the database and receive a Cursor back
        // A cursor is your primary interface to the query results.
        Cursor cursor = db.query(
                MoviesContract.ListEntry.TABLE_NAME,  // Table to Query
                null, // all columns
                null, // Columns for the "where" clause
                null, // Values for the "where" clause
                null, // columns to group by
                null, // columns to filter by row groups
                null // sort order
        );

        assertTrue( "Error: No Records returned from movie query", cursor.moveToFirst() );

        TestUtilities.validateCurrentRecord("Error: Movie Query Validation Failed",
                cursor, list);

        // Move the cursor to demonstrate that there is only one record in the database
        assertFalse( "Error: More than one record returned from movie query",
                cursor.moveToNext() );

        // Close Cursor and Database
        cursor.close();
        db.close();
    }

    @Test
    public void testMovieListTable() {
        MoviesDbHelper dbHelper = new MoviesDbHelper(mContext);
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues list = TestUtilities.createTopRatedListValues();
        Long listId = db.insert(MoviesContract.ListEntry.TABLE_NAME, null, list);
        assertThat(listId, greaterThan(0L));

        ContentValues movie1 = TestUtilities.createReturnOfTheJediValues();
        Long movieId1 = db.insert(MoviesContract.MovieEntry.TABLE_NAME, null, movie1);
        assertThat(movieId1, equalTo(TestUtilities.TEST_MOVIE3_ID));

        ContentValues movie2 = TestUtilities.createStarWarsValues();
        Long movieId2 = db.insert(MoviesContract.MovieEntry.TABLE_NAME, null, movie2);
        assertThat(movieId2, equalTo(TestUtilities.TEST_MOVIE1_ID));

        ContentValues movieList1 = TestUtilities.createMovieListValues(listId, movieId1, 0);
        Long movieListId1 = db.insert(MoviesContract.MovieListEntry.TABLE_NAME, null, movieList1);
        assertThat(movieListId1, greaterThan(0L));

        ContentValues movieList2 = TestUtilities.createMovieListValues(listId, movieId2, 1);
        Long movieListId2 = db.insert(MoviesContract.MovieListEntry.TABLE_NAME, null, movieList2);
        assertThat(movieListId2, greaterThan(0L));

        // Fourth Step: Query the database and receive a Cursor back
        // A cursor is your primary interface to the query results.
        Cursor cursor = db.query(
                MoviesContract.MovieListEntry.TABLE_NAME,  // Table to Query
                null, // all columns
                null, // Columns for the "where" clause
                null, // Values for the "where" clause
                null, // columns to group by
                null, // columns to filter by row groups
                MoviesContract.MovieListEntry.COLUMN_ORDER + " ASC"
        );

        assertTrue( "Error: No Records returned from movie list query", cursor.moveToFirst() );

        TestUtilities.validateCurrentRecord("Error: Movie List Query Validation Failed",
                cursor, movieList1);

        assertTrue( "Error: Just one record returned from movie list query",
                cursor.moveToNext() );

        TestUtilities.validateCurrentRecord("Error: Movie List Query Validation Failed",
                cursor, movieList2);

        assertFalse( "Error: More than two records returned from movie list query",
                cursor.moveToNext() );

        // Close Cursor and Database
        cursor.close();
        db.close();
    }
}

File: app/src/androidTest/java/com/example/popularmovies/data/TestDb.kt
package com.example.popularmovies.data

import androidx.room.Room
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.example.popularmovies.util.MoviesFetcher
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.greaterThan
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

@RunWith(AndroidJUnit4::class)
class TestDb {
    private val mContext = InstrumentationRegistry.getInstrumentation().targetContext

    @Test fun testCreateDb() {
        val tableNameHashSet = HashSet<String>()
        tableNameHashSet.add("movie")
        tableNameHashSet.add("list")
        tableNameHashSet.add("movie_list")

        val db = Room.inMemoryDatabaseBuilder(mContext, MoviesDatabase::class.java).build().openHelper.writableDatabase
        assertEquals(true, db.isOpen)

        var c = db.query("SELECT name FROM sqlite_master WHERE type='table'", // columns to group by
                null)
        assertTrue("Error: This means that the database has not been created correctly",
                c.moveToFirst())

        // verify that the tables have been created
        do {
            tableNameHashSet.remove(c.getString(0))
        } while (c.moveToNext())
        c.close()
        assertTrue("Error: Your database was created without movie, list and movie_list entry tables",
                tableNameHashSet.isEmpty())

        // now, do our tables contain the correct columns?
        c = db.query("PRAGMA table_info(movie)", null)

        assertTrue("Error: Unable to query the database for movie table information.",
                c.moveToFirst())

        // Build a HashSet of all of the column names we want to look for
        val movieColumnHashSet = HashSet<String>()
        movieColumnHashSet.add("_id")
        movieColumnHashSet.add("title")
        movieColumnHashSet.add("overview")
        movieColumnHashSet.add("poster_path")
        movieColumnHashSet.add("release_date")
        movieColumnHashSet.add("adult")
        movieColumnHashSet.add("vote_average")

        var columnNameIndex = c.getColumnIndex("name")
        do {
            val columnName = c.getString(columnNameIndex)
            movieColumnHashSet.remove(columnName)
        } while (c.moveToNext())
        c.close()
        // if this fails, it means that your database doesn't contain all of the required location
        // entry columns
        assertTrue("Error: The database doesn't contain all of the required movie entry columns",
                movieColumnHashSet.isEmpty())

        c = db.query("PRAGMA table_info(list)", null)

        assertTrue("Error: Unable to query the database for list table information.",
                c.moveToFirst())

        // Build a HashSet of all of the column names we want to look for
        val listColumnHashSet = HashSet<String>()
        listColumnHashSet.add("_id")
        listColumnHashSet.add("selection")
        listColumnHashSet.add("date_fetched")

        columnNameIndex = c.getColumnIndex("name")
        do {
            val columnName = c.getString(columnNameIndex)
            listColumnHashSet.remove(columnName)
        } while (c.moveToNext())
        c.close()
        // if this fails, it means that your database doesn't contain all of the required location
        // entry columns
        assertTrue("Error: The database doesn't contain all of the required list entry columns",
                listColumnHashSet.isEmpty())

        c = db.query("PRAGMA table_info(movie_list)", null)

        assertTrue("Error: Unable to query the database for list table information.",
                c.moveToFirst())

        // Build a HashSet of all of the column names we want to look for
        val movieListColumnHashSet = HashSet<String>()
        movieListColumnHashSet.add("_id")
        movieListColumnHashSet.add("list_id")
        movieListColumnHashSet.add("movie_id")
        movieListColumnHashSet.add("order")

        columnNameIndex = c.getColumnIndex("name")
        do {
            val columnName = c.getString(columnNameIndex)
            movieListColumnHashSet.remove(columnName)
        } while (c.moveToNext())
        c.close()
        // if this fails, it means that your database doesn't contain all of the required location
        // entry columns
        assertTrue("Error: The database doesn't contain all of the required movie list entry columns",
                movieListColumnHashSet.isEmpty())

        db.close()
    }

    @Test fun testMovieTable() {
        val db = Room.inMemoryDatabaseBuilder(mContext, MoviesDatabase::class.java).build()

        val movieDao = db.movieDao()
        val movie = TestUtilities.createEmpireStrikesBackValues()
        val movieId: Long?
        movieId = movieDao.insert(movie)
        assertThat(movieId, equalTo(TestUtilities.TEST_MOVIE2_ID))

        val fetchedMovie = movieDao.loadById(TestUtilities.TEST_MOVIE2_ID)

        assertNotNull("Error: No Records returned from movie query", fetchedMovie)

        assertTrue("Error: Movie Query Validation Failed", TestUtilities.compareMovies(movie, fetchedMovie!!))

        db.close()
    }

    @Test fun testListTable() {
        val db = Room.inMemoryDatabaseBuilder(mContext, MoviesDatabase::class.java).build()

        val listDao = db.listDao()
        val list = TestUtilities.createMostPopularListValues()
        val listId: Long?
        listId = listDao.insert(list)
        assertThat(listId, greaterThan(0L))
        list.id = listId

        val fetchedList = listDao.findBySelection(MoviesFetcher.ListSelections.MOST_POPULAR.toString())

        assertNotNull("Error: No Records returned from list query", fetchedList)

        assertTrue("Error: List Query Validation Failed", TestUtilities.compareLists(list, fetchedList!!))

        db.close()
    }

    @Test fun testMovieListTable() {
        val db = Room.inMemoryDatabaseBuilder(mContext, MoviesDatabase::class.java).build()

        val listDao = db.listDao()
        val movieDao = db.movieDao()
        val movieListDao = db.movieListDao()

        val list = TestUtilities.createTopRatedListValues()
        val listId = listDao.insert(list)
        assertThat(listId, greaterThan(0L))

        val movie1 = TestUtilities.createReturnOfTheJediValues()
        val movieId1 = movieDao.insert(movie1)
        assertThat(movieId1, equalTo(TestUtilities.TEST_MOVIE3_ID))

        val movie2 = TestUtilities.createStarWarsValues()
        val movieId2 = movieDao.insert(movie2)
        assertThat(movieId2, equalTo(TestUtilities.TEST_MOVIE1_ID))

        val movieList1 = TestUtilities.createMovieListValues(listId, movieId1, 0)
        val movieListId1 = movieListDao.insert(movieList1)
        assertThat(movieListId1, greaterThan(0L))
        movieList1.id = movieListId1

        val movieList2 = TestUtilities.createMovieListValues(listId, movieId2, 1)
        val movieListId2 = movieListDao.insert(movieList2)
        assertThat(movieListId2, greaterThan(0L))
        movieList2.id = movieListId2

        val movieLists = movieListDao.findByListId(listId)

        assertEquals("Error: No Records returned from movie list query", 2, movieLists.count())

        assertTrue("Error: Movie List Query Validation Failed", TestUtilities.compareMovieLists(movieList1, movieLists[0]))

        assertTrue("Error: Movie List Query Validation Failed", TestUtilities.compareMovieLists(movieList2, movieLists[1]))

        db.close()
    }
}
